### PR TITLE
feat: add progress bar widget

### PR DESCRIPTION
**Pull request by derekdreery**
@PoignardAzur I wanted to have a play with masonry, so I had a go at building a progress bar. I've made a PR in case you want it, but I won't be offended if you close the PR. I'm happy to make changes if you see anything you'd like to change.

### CODE REVIEW

File: masonry/src/widget/progress_bar.rs
Comment (derekdreery) [inline masonry/src/widget/progress_bar.rs:4]: oops this needs updating (there are some other comments as well)

File: masonry/src/widget/progress_bar.rs
Comment (derekdreery) [inline masonry/src/widget/progress_bar.rs:190]: I don't know if I've done this right.

File: masonry/src/widget/progress_bar.rs
Comment (derekdreery) [inline masonry/src/widget/progress_bar.rs:232]: again, not sure if I've done this right.

**Comment by PoignardAzur** [comment]
There's a few things that will need to be tweaked before we can merge this, but it's a great start!

Even then, we probably won't merge this for at least a week or two, until we've implemented most of the pass spec RFC. But after that I'll be super happy to merge this.

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [reply masonry/src/widget/progress_bar.rs:232]: Note that you've added some snapshot/screenshots tests, but you haven't actually added the snapshots and screenshots yet, so the tests will necessary fail.

Also, you can probably add a few more screenshot tests for different levels of completion in the progress bar. (I'd say 0%, 50%, 100% and None are the ones you're interested in.)

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [reply masonry/src/widget/progress_bar.rs:190]: Judging from how egui does it, you want something like:

```suggestion
        ctx.current_node().set_value(self.value_accessibility());
        if let Some(value) = self.part_complete {
            ctx.current_node().set_value(value * 100.0);
        }
```

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:27]: I'd rename this (and the matching methods and function arguments) to `progress` instead.

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:198]:
```suggestion
        trace_span!("ProgressBar")
```

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:203]: I like that!

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:214]: This is a huge nitpick, but this should probably be moved above the Widget impl.

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:233]:
```suggestion
        assert_render_snapshot!(harness, "default_bar");
```

File: xilem/src/view/progress_bar.rs
Comment (PoignardAzur) [inline xilem/src/view/progress_bar.rs:14]: Same deal, this should be renamed to `progress`.

File: xilem/src/view/progress_bar.rs
Comment (PoignardAzur) [inline xilem/src/view/progress_bar.rs:68]: You can probably remove most of this. Progress bars don't have user interactions.

```suggestion
        tracing::error!("Message arrived in ProgressBar::ProgressBar, but Label doesn't consume any messages, this is a bug");
        MessageResult::Stale(message)
```

**Review by PoignardAzur — CHANGES_REQUESTED** [review]
Super happy to see new widget contributions.

File: masonry/src/widget/progress_bar.rs
Comment (derekdreery) [reply masonry/src/widget/progress_bar.rs]: I tried this code, but the `set_value` function expects a string. I think the existing code already reports the value in an accessibility-friendly way.

File: masonry/src/widget/progress_bar.rs
Comment (waywardmonkeys) [inline masonry/src/widget/progress_bar.rs:7]: Could you import this below where you already import `Size`?

**Review by waywardmonkeys — COMMENTED** [review]
Would you be open to naming it `Progress` instead of `ProgressBar` since not all progress displays would be bars:

Some examples from a library that I've enjoyed using: https://ant.design/components/progress

**Comment by PoignardAzur** [comment]
This looks great! Don't really have the energy to review right now, will try this weekend.

File: xilem/examples/widgets.rs
Comment (PoignardAzur) [inline xilem/examples/widgets.rs:94]: Wait, does that fix the Cargo warning we keep getting?

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:215]: Sorry, I made a typo last time. I meant [set_numeric_value](https://docs.rs/accesskit/latest/accesskit/struct.NodeBuilder.html#method.set_numeric_value).

```suggestion
        ctx.current_node().set_value(self.value_accessibility());
        if let Some(value) = self.progress {
            ctx.current_node().set_numeric_value(value * 100.0);
        }
```

**Review by PoignardAzur — APPROVED** [review]
LGTM.

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:203]: Sorry, I made a typo last time. I meant [set_numeric_value](https://docs.rs/accesskit/latest/accesskit/struct.NodeBuilder.html#method.set_numeric_value).

```suggestion
        ctx.current_node().set_value(self.value_accessibility());
        if let Some(value) = self.progress {
            ctx.current_node().set_numeric_value(value * 100.0);
        }
```

**Comment by PoignardAzur** [comment]
Anyone know why cargo clippy fails in CI?

**Comment by waywardmonkeys** [comment]
> Anyone know why cargo clippy fails in CI?

Beyond that there are 3 clippy errors?

**Comment by Philipp-M** [comment]
To expand (well really). click on the arrow above the error.

File: xilem/examples/widgets.rs
Comment (DJMcNab) [reply xilem/examples/widgets.rs:94]: This fixes an error which occurs if you try and compile the not `_android` example for Android (e.g. using `cargo apk check --example mason -p android`)

This is mostly a workaround for if you set up your development environment for android compilation (see e.g. https://github.com/linebender/vello/blob/a00cd9ffb617810980487b88a6bfa13d3a642e40/.vscode/settings.json#L23-L36)

**Comment by PoignardAzur** [comment]
Oh, that UI is obtuse. Thanks.

**Comment by PoignardAzur** [comment]
(Quick reminder: if this PR can be rebased and the clippy errors addressed, this is ready to merge.)

**Comment by waywardmonkeys** [comment]
I still think it should end up being `Progress` as mentioned above.

**Comment by PoignardAzur** [comment]
For now I'd say that doesn't matter. We can rename it to `Progress` or `ProgressIndicator` or something if we implement a circular styling option.

**Comment by derekdreery** [comment]
Aah I thought it would be longer before I could merge. Will aim to rebase ASAP.

File: masonry/src/widget/progress_bar.rs
Comment (jaredoconnell) [inline masonry/src/widget/progress_bar.rs:27]: Why did you choose a f32 for this? The precedent set is ctx::set_numeric_value, which takes an f64 as of the current main code. I changed this to f64 when I rebased this to main. Do you want me to push that change?

**Review by jaredoconnell — COMMENTED** [review]
I merged main into this branch to test it. That fixes a lagginess that I was experiencing before. Do you want me to push that change along with the switch to f64?

File: masonry/src/widget/progress_bar.rs
Comment (derekdreery) [reply masonry/src/widget/progress_bar.rs:27]: yes feel free

**Comment by jaredoconnell** [comment]
I rebased it to main. It works pretty well. Since I made some changes, I'll stay out of reviewing. All that remains is a clippy error due to an empty body for the access events. What do we want to do about that? Should we add a clippy allow (`#[allow(...)]`) directive to those lines?

File: masonry/src/widget/progress_bar.rs
Comment (PoignardAzur) [inline masonry/src/widget/progress_bar.rs:121]:
```suggestion
    // pointer events unhandled for now
    fn on_pointer_event(&mut self, _ctx: &mut EventCtx, _event: &PointerEvent) {}

    fn on_text_event(&mut self, _ctx: &mut EventCtx, _event: &TextEvent) {}

    // access events unhandled for now
    fn on_access_event(&mut self, _ctx: &mut EventCtx, _event: &AccessEvent) {}
```

File: xilem/src/view/progress_bar.rs
Comment (PoignardAzur) [inline xilem/src/view/progress_bar.rs:56]:
```suggestion
        tracing::error!("Message arrived in ProgressBar::message, but ProgressBar doesn't consume any messages, this is a bug");
```

**Review by PoignardAzur — APPROVED** [review]
LGTM.

**Review by jaredoconnell — APPROVED** [review]
This looks ready to merge.

File: masonry/src/widget/progress_bar.rs
Comment (DJMcNab) [inline masonry/src/widget/progress_bar.rs:105]: for now? What would you expect a pointer event on a progress bar to do?

Maybe copy the current progress percentage to the clipboard?

File: xilem/src/view/progress_bar.rs
Comment (DJMcNab) [inline xilem/src/view/progress_bar.rs:23]: Why is this an action widget? The progressbar widget doesn't seem to send actions as far as I can see

**Review by DJMcNab — COMMENTED** [review]
A couple clarifications I'd like to see

**Pull request merged**
